### PR TITLE
Playerbot functionnality to use (channels / emote / say / yell / guild) chat

### PR DESCRIPTION
**Pull request by atidot3**
tested out with 5000 active bots havent seen any lag / crash but it spam alot.

### CODE REVIEW

**Comment by atidot3** [comment]
PR need core changes:
https://github.com/liyunfan1223/azerothcore-wotlk/pull/15

File: src/AiFactory.cpp [hunk 1]
```diff
@@ -439,26 +439,28 @@ void AiFactory::AddDefaultCombatStrategies(Player* player, PlayerbotAI* const fa
...
 
-        engine->addStrategies("boost", "racials", "chat", "default", "aoe", "potions", "cast time", "dps assist", nullptr);
+        engine->addStrategies("boost", "racials", "chat", "default", "aoe", "potions", "cast time", "dps assist", "pvp", nullptr);
         engine->removeStrategy("custom::say");
         engine->removeStrategy("flee");
         engine->removeStrategy("threat");
-        engine->addStrategy("boost");
+        engine->removeStrategy("follow");
+        engine->removeStrategy("conserve mana");
+        engine->removeStrategy("cast time");
```
Comment (liyunfan1223) [inline src/AiFactory.cpp]: Don’t just change it for granted, you need to make sure the strategies are working correctly and explain what specific problem they solve.

File: src/AiFactory.cpp [hunk 2]
```diff
@@ -674,10 +682,10 @@ void AiFactory::AddDefaultNonCombatStrategies(Player* player, PlayerbotAI* const
 
 Engine* AiFactory::createNonCombatEngine(Player* player, PlayerbotAI* const facade, AiObjectContext* aiObjectContext)
 {
-	Engine* nonCombatEngine = new Engine(facade, aiObjectContext);
+    Engine* nonCombatEngine = new Engine(facade, aiObjectContext);
```
Comment (liyunfan1223) [inline src/AiFactory.cpp]: Don't adjust the code style in this PR, we need to focus on the feature you introduced.

File: src/PlayerbotMgr.cpp [hunk 3]
```diff
@@ -85,7 +85,8 @@ void PlayerbotHolder::HandlePlayerBotLoginCallback(PlayerbotLoginQueryHolder con
 
     uint32 botAccountId = holder.GetAccountId();
 
-    WorldSession* botSession = new WorldSession(botAccountId, "", nullptr, SEC_PLAYER, EXPANSION_WRATH_OF_THE_LICH_KING, time_t(0), LOCALE_enUS, 0, false, false, 0, true);
+    // At login DBC locale should be what the server is set to use by default (as spells etc are hardcoded to ENUS this allows channels to work as intended)
+    WorldSession* botSession = new WorldSession(botAccountId, "", nullptr, SEC_PLAYER, EXPANSION_WRATH_OF_THE_LICH_KING, time_t(0), sWorld->GetDefaultDbcLocale(), 0, false, false, 0, true);
```
Comment (liyunfan1223) [inline src/PlayerbotMgr.cpp:89]: After the change, you should modify the calculation of spell id value

https://github.com/liyunfan1223/mod-playerbots/blob/8816758490a0e33327f53af2c31d1f369044be5a/src/strategy/values/SpellIdValue.cpp#L193

to `LocaleConstant loc = LOCALE_enUS;`

File: src/AiFactory.cpp [hunk 4]
```diff
@@ -631,15 +638,16 @@ void AiFactory::AddDefaultNonCombatStrategies(Player* player, PlayerbotAI* const
...
     // Battleground switch
     if (player->InBattleground() && player->GetBattleground())
     {
-        nonCombatEngine->addStrategies("nc", "chat", "default", "buff", "food", "mount", "pvp", "dps assist", "attack tagged", nullptr);
+        nonCombatEngine->addStrategies("racials", "nc", "chat", "default", "buff", "food", "mount", "collision", "pvp", "dps assist", "attack tagged", "emote", nullptr);
         nonCombatEngine->removeStrategy("custom::say");
         nonCombatEngine->removeStrategy("travel");
         nonCombatEngine->removeStrategy("rpg");
+        nonCombatEngine->removeStrategy("rpg craft");
+        nonCombatEngine->removeStrategy("follow");
```
Comment (liyunfan1223) [inline src/AiFactory.cpp:650]: Don’t just change it for granted, you need to make sure the strategies are working correctly and explain what specific problem they solve.

**Review by liyunfan1223 — CHANGES_REQUESTED** [review]
Good work, please make some changes as required.

It looks like we need to test with different dbc files on different servers. And it's still in the debug phase, it's best to set a config to control the feature on/off (default is off).

File: src/AiFactory.cpp [hunk 5]
```diff
@@ -439,26 +439,28 @@ void AiFactory::AddDefaultCombatStrategies(Player* player, PlayerbotAI* const fa
...
         if ((player->getClass() == CLASS_DRUID && tab == 2) || (player->getClass() == CLASS_SHAMAN && tab == 2))
             engine->addStrategies("caster", "caster aoe", nullptr);
 
         if (player->getClass() == CLASS_DRUID && tab == 1)
-            engine->addStrategies(/*"behind",*/ "dps", nullptr);
+            engine->addStrategies("behind", "dps", nullptr);
 
         if (player->getClass() == CLASS_ROGUE)
-            engine->addStrategies(/*"behind",*/ "stealth", nullptr);
+            engine->addStrategies("behind", "stealth", nullptr);
```
Comment (liyunfan1223) [inline src/AiFactory.cpp]: Don’t just change it for granted, you need to make sure the strategies are working correctly and explain what specific problem they solve.

File: src/AiFactory.cpp (hunk 1 above)
Comment (atidot3) [reply src/AiFactory.cpp]: yea forgot to switch bvranch ... sry

File: src/strategy/Engine.cpp [hunk 6]
```diff
@@ -172,6 +172,7 @@ bool Engine::DoNextAction(Unit* unit, uint32 depth, bool minimal)
 
             if (!action)
             {
+                LOG_ERROR("playerbots", "Action: {} - is UNKNOWN - c:{} l:{}", actionNode->getName().c_str(), botAI->GetBot()->getClass(), botAI->GetBot()->GetLevel());
```
Comment (liyunfan1223) [inline src/strategy/Engine.cpp]: remove this error log

**Review by liyunfan1223 — CHANGES_REQUESTED** [review]
I did a test myself, based on a commit to https://github.com/atidot3/mod-playerbots/commit/5f60ded4198a7eba57ceb59f74e0619109166820. Overall it works fine, but there are a few issues present

1. With EnableGreet or RandomBotEmote enabled, grouping a number of bots (like a 25-man team) and summoning them together produces excessive greeting and emote content.

2. They actively talk in the world channel, but I don't get messages in other channels (like trade or lfg).

We can merge it first after remove this error log. Other issues can be resolved after merging. It would be great if more people were involved in testing and developing this feature and giving feedback.

**Review by liyunfan1223 — APPROVED** [review]

**Pull request merged**